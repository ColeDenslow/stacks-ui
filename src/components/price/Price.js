import React from 'react';
import cn from 'classnames';
import PropTypes from 'prop-types';
import { formatMoneyCustomized, formatMoneyByInternationalizationStandard } from './moneyHelper';

import './Price.css';

function Price(props) {
  const { amount, amountBeforeDiscount, amountLabel, amountBeforeDiscountLabel, className, intlMoneyFormatObj,
    moneyFormatString, textAlign, ...otherProps } = props;

  const getFormattedPrice = function (price) {
    if (moneyFormatString) {
      return formatMoneyCustomized(price, moneyFormatString);
    } else {
      return formatMoneyByInternationalizationStandard(price, intlMoneyFormatObj);
    }
  };

  if (amount !== null) {
    const priceClassNames = cn([
      'stx-price',
      className,
      { 'stx-price--alight-left': textAlign === 'left' },
      { 'stx-price--alight-center': textAlign === 'center' },
      { 'stx-price--alight-right': textAlign === 'right' },
    ]);

    const amountClassNames = cn([
      'stx-price__amount',
      { 'stx-price__amount--inactive': amountBeforeDiscount !== null },
    ]);

    return (
      <div
        className={priceClassNames}
        {...otherProps}
      >
        { amountBeforeDiscount !== null &&
          <p className={amountClassNames} aria-label={amountBeforeDiscountLabel}>
            { getFormattedPrice(amountBeforeDiscount) }
          </p>
        }
        <p className='stx-price__amount' aria-label={amountLabel}>
          { getFormattedPrice(amount) }
        </p>
      </div>
    );
  }
  return null;
}

Price.defaultProps = {
  className: '',
  textAlign: 'center',
  amount: null,
  amountBeforeDiscount: null,
  amountLabel: 'Product price',
  amountBeforeDiscountLabel: 'Price before discount',
  moneyFormatString: null,
  intlMoneyFormatObj: { languageTag: 'en-CA', currency: 'CAD' },
};

Price.propTypes = {
  /** Value appended to the class attribute */
  className: PropTypes.string,

  /** How the price text should be aligned */
  textAlign: PropTypes.oneOf(['left', 'center', 'right']),

  /** Value of the price amount (cents), for example: $24.99 would be 2499. */
  amount: PropTypes.number,

  /** Description for the price component, used by screen readers (accessibility purposes). */
  amountLabel: PropTypes.string,

  /** Value of the discounted price amount (cents), for example: $24.99 should be passed as 2499. */
  amountBeforeDiscount: PropTypes.number,

  /** Description for the component that shows price before discount, used by screen readers (accessibility purposes). */
  amountBeforeDiscountLabel: PropTypes.string,

  /** A formatting string, examples: 'CA${{amount}}', 'â‚¬{{amount_with_comma_separator}}' */
  moneyFormatString: PropTypes.string,

  /** The formatting object */
  intlMoneyFormatObj: PropTypes.object,
};

export default Price;
