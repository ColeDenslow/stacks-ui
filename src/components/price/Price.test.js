import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Price from './Price';

/* eslint-disable no-template-curly-in-string */

describe('Price', () => {
  test('Price renders with basic props', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={1222}
        amountBeforeDiscount={977}
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price renders with zero amount', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={0}
        amountBeforeDiscount={977}
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price not renders with null amount', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={null}
        amountBeforeDiscount={977}
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price renders with intlMoneyFormatObj and no amount before discount', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={1222}
        intlMoneyFormatObj={{
          languageTag: 'en-CA',
          currency: 'CAD',
        }}
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price renders with intlMoneyFormatObj for foreign language/currency that uses no fractional unit', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={127611}
        intlMoneyFormatObj={{
          languageTag: 'ja-JP',
          currency: 'JPY',
        }}
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price renders with intlMoneyFormatObj and textAlign left', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={1222}
        textAlign='left'
        intlMoneyFormatObj={{
          languageTag: 'en-CA',
          currency: 'CAD',
        }}
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price renders with intlMoneyFormatObj and textAlign right', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={1222}
        textAlign='right'
        intlMoneyFormatObj={{
          languageTag: 'en-CA',
          currency: 'CAD',
        }}
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price renders with intlMoneyFormatObj and basic props', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={1222}
        amountBeforeDiscount={977}
        intlMoneyFormatObj={{
          languageTag: 'en-CA',
          currency: 'CAD',
        }}
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price renders with formatString and basic props', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={1222}
        amountBeforeDiscount={977}
        moneyFormatString='${{amount}}'
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price renders with formatString and customized currency', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={149999}
        amountBeforeDiscount={175000}
        moneyFormatString='US${{amount}}'
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price renders with formatString and customized currency/separators', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={149999}
        amountBeforeDiscount={175000}
        moneyFormatString='€{{amount_with_comma_separator}}'
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price renders with formatString and no cents', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={1499}
        amountBeforeDiscount={1750}
        moneyFormatString='${{amount_no_decimals}}'
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price renders with formatString and no cents + dot as thousands separator', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={149999}
        amountBeforeDiscount={175000}
        moneyFormatString='€{{amount_no_decimals_with_comma_separator}}'
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price renders with formatString and space separator', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={149999}
        amountBeforeDiscount={175000}
        moneyFormatString='${{amount_with_space_separator}}'
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price renders with formatString and apostrophe separator', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={149999}
        amountBeforeDiscount={175000}
        moneyFormatString='${{amount_with_apostrophe_separator}}'
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });

  test('Price renders with customized/accessibility props', () => {
    const output = mount(
      <Price
        className='the-class'
        amount={1222}
        amountBeforeDiscount={977}
        aria-label='product-price'
        data-additional-prop='additional-prop'
      />
    );

    expect(toJson(output)).toMatchSnapshot();
  });
});
