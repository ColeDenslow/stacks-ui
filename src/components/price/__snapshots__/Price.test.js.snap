// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Price Price not renders with null amount 1`] = `
<Price
  amount={null}
  amountBeforeDiscount={977}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString={null}
  textAlign="center"
/>
`;

exports[`Price Price renders with basic props 1`] = `
<Price
  amount={1222}
  amountBeforeDiscount={977}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString={null}
  textAlign="center"
>
  <div
    className="stx-price the-class stx-price--alight-center"
  >
    <p
      aria-label="Price before discount"
      className="stx-price__amount stx-price__amount--inactive"
    >
      $9.77
    </p>
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      $12.22
    </p>
  </div>
</Price>
`;

exports[`Price Price renders with customized/accessibility props 1`] = `
<Price
  amount={1222}
  amountBeforeDiscount={977}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  aria-label="product-price"
  className="the-class"
  data-additional-prop="additional-prop"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString={null}
  textAlign="center"
>
  <div
    aria-label="product-price"
    className="stx-price the-class stx-price--alight-center"
    data-additional-prop="additional-prop"
  >
    <p
      aria-label="Price before discount"
      className="stx-price__amount stx-price__amount--inactive"
    >
      $9.77
    </p>
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      $12.22
    </p>
  </div>
</Price>
`;

exports[`Price Price renders with formatString and apostrophe separator 1`] = `
<Price
  amount={149999}
  amountBeforeDiscount={175000}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString="\${{amount_with_apostrophe_separator}}"
  textAlign="center"
>
  <div
    className="stx-price the-class stx-price--alight-center"
  >
    <p
      aria-label="Price before discount"
      className="stx-price__amount stx-price__amount--inactive"
    >
      $1'750.00
    </p>
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      $1'499.99
    </p>
  </div>
</Price>
`;

exports[`Price Price renders with formatString and basic props 1`] = `
<Price
  amount={1222}
  amountBeforeDiscount={977}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString="\${{amount}}"
  textAlign="center"
>
  <div
    className="stx-price the-class stx-price--alight-center"
  >
    <p
      aria-label="Price before discount"
      className="stx-price__amount stx-price__amount--inactive"
    >
      $9.77
    </p>
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      $12.22
    </p>
  </div>
</Price>
`;

exports[`Price Price renders with formatString and customized currency 1`] = `
<Price
  amount={149999}
  amountBeforeDiscount={175000}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString="US\${{amount}}"
  textAlign="center"
>
  <div
    className="stx-price the-class stx-price--alight-center"
  >
    <p
      aria-label="Price before discount"
      className="stx-price__amount stx-price__amount--inactive"
    >
      US$1,750.00
    </p>
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      US$1,499.99
    </p>
  </div>
</Price>
`;

exports[`Price Price renders with formatString and customized currency/separators 1`] = `
<Price
  amount={149999}
  amountBeforeDiscount={175000}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString="€{{amount_with_comma_separator}}"
  textAlign="center"
>
  <div
    className="stx-price the-class stx-price--alight-center"
  >
    <p
      aria-label="Price before discount"
      className="stx-price__amount stx-price__amount--inactive"
    >
      €1.750,00
    </p>
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      €1.499,99
    </p>
  </div>
</Price>
`;

exports[`Price Price renders with formatString and no cents + dot as thousands separator 1`] = `
<Price
  amount={149999}
  amountBeforeDiscount={175000}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString="€{{amount_no_decimals_with_comma_separator}}"
  textAlign="center"
>
  <div
    className="stx-price the-class stx-price--alight-center"
  >
    <p
      aria-label="Price before discount"
      className="stx-price__amount stx-price__amount--inactive"
    >
      €1.750
    </p>
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      €1.500
    </p>
  </div>
</Price>
`;

exports[`Price Price renders with formatString and no cents 1`] = `
<Price
  amount={1499}
  amountBeforeDiscount={1750}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString="\${{amount_no_decimals}}"
  textAlign="center"
>
  <div
    className="stx-price the-class stx-price--alight-center"
  >
    <p
      aria-label="Price before discount"
      className="stx-price__amount stx-price__amount--inactive"
    >
      $18
    </p>
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      $15
    </p>
  </div>
</Price>
`;

exports[`Price Price renders with formatString and space separator 1`] = `
<Price
  amount={149999}
  amountBeforeDiscount={175000}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString="\${{amount_with_space_separator}}"
  textAlign="center"
>
  <div
    className="stx-price the-class stx-price--alight-center"
  >
    <p
      aria-label="Price before discount"
      className="stx-price__amount stx-price__amount--inactive"
    >
      $1 750,00
    </p>
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      $1 499,99
    </p>
  </div>
</Price>
`;

exports[`Price Price renders with intlMoneyFormatObj and basic props 1`] = `
<Price
  amount={1222}
  amountBeforeDiscount={977}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString={null}
  textAlign="center"
>
  <div
    className="stx-price the-class stx-price--alight-center"
  >
    <p
      aria-label="Price before discount"
      className="stx-price__amount stx-price__amount--inactive"
    >
      $9.77
    </p>
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      $12.22
    </p>
  </div>
</Price>
`;

exports[`Price Price renders with intlMoneyFormatObj and no amount before discount 1`] = `
<Price
  amount={1222}
  amountBeforeDiscount={null}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString={null}
  textAlign="center"
>
  <div
    className="stx-price the-class stx-price--alight-center"
  >
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      $12.22
    </p>
  </div>
</Price>
`;

exports[`Price Price renders with intlMoneyFormatObj and textAlign left 1`] = `
<Price
  amount={1222}
  amountBeforeDiscount={null}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString={null}
  textAlign="left"
>
  <div
    className="stx-price the-class stx-price--alight-left"
  >
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      $12.22
    </p>
  </div>
</Price>
`;

exports[`Price Price renders with intlMoneyFormatObj and textAlign right 1`] = `
<Price
  amount={1222}
  amountBeforeDiscount={null}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString={null}
  textAlign="right"
>
  <div
    className="stx-price the-class stx-price--alight-right"
  >
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      $12.22
    </p>
  </div>
</Price>
`;

exports[`Price Price renders with intlMoneyFormatObj for foreign language/currency that uses no fractional unit 1`] = `
<Price
  amount={127611}
  amountBeforeDiscount={null}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "JPY",
      "languageTag": "ja-JP",
    }
  }
  moneyFormatString={null}
  textAlign="center"
>
  <div
    className="stx-price the-class stx-price--alight-center"
  >
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      ￥1,276
    </p>
  </div>
</Price>
`;

exports[`Price Price renders with zero amount 1`] = `
<Price
  amount={0}
  amountBeforeDiscount={977}
  amountBeforeDiscountLabel="Price before discount"
  amountLabel="Product price"
  className="the-class"
  intlMoneyFormatObj={
    Object {
      "currency": "CAD",
      "languageTag": "en-CA",
    }
  }
  moneyFormatString={null}
  textAlign="center"
>
  <div
    className="stx-price the-class stx-price--alight-center"
  >
    <p
      aria-label="Price before discount"
      className="stx-price__amount stx-price__amount--inactive"
    >
      $9.77
    </p>
    <p
      aria-label="Product price"
      className="stx-price__amount"
    >
      $0.00
    </p>
  </div>
</Price>
`;
